client/blocks/site-preview/index.jsx

```diff
@@ -1,110 +1,82 @@
-	getPreviewSiteId,
-	getPreviewUrl,
-} from 'calypso/state/ui/preview/selectors';
+import { getPreviewSiteId, getPreviewUrl } from 'calypso/state/ui/preview/selectors';
 
-const debug = debugFactory( 'calypso:site-preview' );
+function usePreviewCounter( siteId, showPreview ) {
+	const [ previewCount, setPreviewCount ] = useState( 0 );
 
-class SitePreview extends Component {
-	static propTypes = {
-		className: PropTypes.string,
-		showPreview: PropTypes.bool,
-		previewUrl: PropTypes.string,
-		selectedSite: PropTypes.object,
-		selectedSiteId: PropTypes.number,
-		selectedSiteNonce: PropTypes.string,
-		selectedSiteUrl: PropTypes.string,
-		closePreview: PropTypes.func.isRequired,
-	};
+	useEffect( () => {
+		if ( siteId ) {
+			setPreviewCount( 0 );
+		}
+	}, [ siteId ] );
 
-	state = {
-		previewCount: 0,
-	};
+	useEffect( () => {
+		if ( showPreview ) {
+			setPreviewCount( ( n ) => n + 1 );
+		}
+	}, [ showPreview ] );
 
-	previewCounter = 0;
+	return previewCount;
+}
 
-	// @TODO: Please update https://github.com/Automattic/wp-calypso/issues/58453 if you are refactoring away from UNSAFE_* lifecycle methods!
-	UNSAFE_componentWillReceiveProps( nextProps ) {
-		if ( this.props.selectedSiteId && this.props.selectedSiteId !== nextProps.selectedSiteId ) {
-			this.previewCounter = 0;
-		}
+function SitePreview( { className } ) {
+	const dispatch = useDispatch();
 
-		if ( ! this.props.showPreview && nextProps.showPreview ) {
-			debug( 'forcing refresh' );
-			this.previewCounter > 0 && this.setState( { previewCount: this.previewCounter } );
-			this.previewCounter += 1;
-		}
+	const selectedSiteId = useSelector( getPreviewSiteId );
+	const siteSlug = useSelector( ( state ) => getSiteSlug( state, selectedSiteId ) );
+	const selectedSiteUrl = `https://${ siteSlug }`.replace( /::/g, '/' );
+	const showPreview = useSelector( ( state ) => getCurrentLayoutFocus( state ) === 'preview' );
+	const selectedSiteNonce = useSelector( ( state ) =>
+		getSiteOption( state, selectedSiteId, 'frame_nonce' )
+	);
+	const selectedSitePreviewable = useSelector( ( state ) =>
+		isSitePreviewable( state, selectedSiteId )
+	);
+	const previewUrl = useSelector( getPreviewUrl );
+	const hideSEO = useSelector( ( state ) => isDomainOnlySite( state, selectedSiteId ) );
+
+	const previewCount = usePreviewCounter( selectedSiteId, showPreview );
+	const basePreviewUrl = previewUrl || selectedSiteUrl;
+
+	if ( ! selectedSitePreviewable ) {
+		return null;
 	}
 
-	getPreviewUrl() {
-		if ( ! this.props.selectedSiteUrl && ! this.props.previewUrl ) {
-			debug( 'no preview url and no site url were found for this site' );
+	function formatPreviewUrl() {
+		if ( ! selectedSiteUrl && ! previewUrl ) {
 			return null;
 		}
-		const previewUrl = addQueryArgs(
+
+		return addQueryArgs(
 			{
 				iframe: true,
 				theme_preview: true,
-				'frame-nonce': this.props.selectedSiteNonce,
-				cachebust: this.state.previewCount,
+				'frame-nonce': selectedSiteNonce ?? '',
```

Comment:
Because the `siteId` and `previewUrl` are set separately and non-atomically:
```js
this.props.setAllSitesPreviewSiteId( siteId );
this.props.setPreviewUrl( previewUrl );
```
we render twice: old preview URL with new site's nonce (bad!), and new preview URL with new site's nonce (good).

To improve this, could be merge both actions into one, with optional `siteId` parameter?
```js
this.props.setPreviewUrl( previewUrl, siteId );
```
If `siteId` is omitted, the reducer will reset `currentPreviewSiteId` to `null`.

Reply:
I reworked the `previewCount` implementation in https://github.com/Automattic/wp-calypso/pull/63008/commits/8664eaf4a7e5bb6e2d44c64143b98a4f526e601c which should cover with the suggestions you made. Let me know what you think.